README.md

### ![#f03c15](https://placehold.it/15/f03c15/000000?text=+) Community Announcements ![#f03c15](https://placehold.it/15/f03c15/000000?text=+)

* **Oct 31 2020** [Cats 2.3.0-M1 is released](https://github.com/typelevel/cats/releases/tag/v2.3.0-M1) with Dotty supported
* **Sept 3 2020** [Cats 2.2.0 is released](https://github.com/typelevel/cats/releases/tag/v2.2.0)
* **Dec 18 2019** [Cats 2.1.0 is released](https://github.com/typelevel/cats/releases/tag/v2.1.0)
* **Sep 9 2019** [Cats 2.0.0 is released](https://github.com/typelevel/cats/releases/tag/v2.0.0)
* **Jun 3 2019** [Cats 1.6.1 is released](https://github.com/typelevel/cats/releases/tag/v1.6.1) with backported bug fixes
* **Apr 26 2019** [We launched a sustainability program](https://typelevel.org/blog/2019/04/24/typelevel-sustainability-program-announcement.html). Please consider supporting us.

## Cats

![cats image](http://plastic-idolatry.com/erik/cats2.png)

![Continuous Integration](https://github.com/typelevel/cats/workflows/Continuous%20Integration/badge.svg)
[![Financial Contributors on Open Collective](https://opencollective.com/typelevel/all/badge.svg?label=financial+contributors)](https://opencollective.com/typelevel) [![Chat](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/typelevel/cats)
[![Latest version](https://index.scala-lang.org/typelevel/cats/cats-core/latest.svg?color=orange&v=1)](https://index.scala-lang.org/typelevel/cats/cats-core)
[![Scala.js](http://scala-js.org/assets/badges/scalajs-1.2.0.svg)](http://scala-js.org)


### Overview

Cats is a library which provides abstractions for functional programming in the [Scala programming language](https://scala-lang.org).
The name is a playful shortening of the word *category*.

Scala supports both object-oriented and functional programming, and this is reflected in the hybrid approach of the
standard library. Cats strives to provide functional programming abstractions that are core, [binary compatible](http://typelevel.org/cats/#binary-compatibility-and-versioning), [modular](http://typelevel.org/cats/motivations#modularity), [approachable](http://typelevel.org/cats/motivations#approachability) and [efficient](http://typelevel.org//cats/motivations#efficiency). A broader goal of Cats is to provide a foundation for an [ecosystem of pure, typeful libraries](https://typelevel.org/cats/#ecosystem) to support functional programming in Scala applications.

For more detail about Cats' motivations, go [here](http://typelevel.org/cats/motivations).

## Contributors

### Code Contributors

This project exists thanks to [all the people who contribute](https://github.com/typelevel/cats/graphs/contributors). We welcome contributions to Cats and would love for you to help build
Cats. See our [contributor guide](https://typelevel.org/cats/contributing.html) for more
information about how you can get involved as a developer. If you are looking for something to start with, [here is a beginner friendly list](https://github.com/typelevel/cats/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22). 

### Financial Contributors

[Become a financial contributor](https://opencollective.com/typelevel) and help us sustain our community. Donations directly support office hour for maintainers, better documentation and strategic initiatives. 



### Gold Sponsors 

Gold Sponsors are those who have pledged $5,000 to $10,000.

<a href="http://47deg.com"><img src="http://typelevel.org/cats/img/sponsors/47_degree.png" style="margin-left: 10px; margin-right: 10px" /></a>
<a href="https://www.iteratorshq.com"><img src="http://typelevel.org/cats/img/sponsors/iterators.png" style="margin-bottom:8px;margin-left: 10px; margin-right: 10px"/></a>
<a href="https://triplequote.com/"><img src="http://typelevel.org/cats/img/sponsors/triplequote.png" style="margin-bottom:9px; margin-right: 10px" /></a>
<a href="http://underscore.io"><img src="http://typelevel.org/cats/img/sponsors/underscore.png" style="margin-bottom:5px;margin-left: 10px; margin-right: 10px"/></a>



### Silver Sponsors

Silver Sponsors are those who have pledged $2,000 to $5,000.

<a href="https://www.ebiznext.com/"><img src="http://typelevel.org/cats/img/sponsors/ebiznext.png" style="margin-bottom:15px; margin-right: 10px" /></a>
<a href="https://www.inner-product.com/"><img src="http://typelevel.org/cats/img/sponsors/inner-product.png" style="margin-bottom:10px; margin-right: 10px"/></a>
<a href="https://evolutiongaming.com/"><img src="http://typelevel.org/cats/img/sponsors/evolution_gaming_engineering.png" style="margin-bottom:10px; margin-right: 10px"/></a>


#### Other Organizations

Support this project with your organization. Your logo will show up here with a link to your website. [[Contribute](https://opencollective.com/typelevel/contribute)]

<a href="https://opencollective.com/typelevel/organization/0/website"><img src="https://opencollective.com/typelevel/organization/0/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/1/website"><img src="https://opencollective.com/typelevel/organization/1/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/2/website"><img src="https://opencollective.com/typelevel/organization/2/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/3/website"><img src="https://opencollective.com/typelevel/organization/3/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/4/website"><img src="https://opencollective.com/typelevel/organization/4/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/5/website"><img src="https://opencollective.com/typelevel/organization/5/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/6/website"><img src="https://opencollective.com/typelevel/organization/6/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/7/website"><img src="https://opencollective.com/typelevel/organization/7/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/8/website"><img src="https://opencollective.com/typelevel/organization/8/avatar.svg"></a>
<a href="https://opencollective.com/typelevel/organization/9/website"><img src="https://opencollective.com/typelevel/organization/9/avatar.svg"></a>


#### Individuals

<a href="https://opencollective.com/typelevel"><img src="https://opencollective.com/typelevel/individuals.svg?width=890"></a>



### Getting Started

Cats is currently available for Scala 2.10 (up to 1.2.x), 2.11, 2.12, 2.13, [Scala.js](http://www.scala-js.org/), and [Scala Native](https://www.scala-native.org/).

Cats relies on improved type inference via the fix for [SI-2712](https://github.com/scala/bug/issues/2712), which is not enabled by default. For **Scala 2.11.9+ or 2.12** you should add the following to your `build.sbt`:

```scala
scalacOptions += "-Ypartial-unification"
```

(Partial unification is on by default since Scala 2.13, the compiler no longer accepts `-Ypartial-unification`)

**Or**, if you need to support older versions of Scala you can use the [sbt-partial-unification](https://github.com/fiadliel/sbt-partial-unification#sbt-partial-unification) plugin which extends support back through **Scala 2.10.6 or later**, to add it, simply add this line to your `plugins.sbt`:

```scala
addSbtPlugin("org.lyranthe.sbt" % "partial-unification" % "1.1.2")
```

And then create the Cats dependency, by adding the following to your `build.sbt`:

```scala
libraryDependencies += "org.typelevel" %% "cats-core" % "2.3.0"
```

This will pull in the cats-core module. If you require some other
functionality, you can pick-and-choose from amongst these modules
(used in place of `"cats-core"`):

 * `cats-kernel`: Small set of basic type classes (*required*).
 * `cats-core`: Most core type classes and functionality (*required*).
 * `cats-laws`: Laws for testing type class instances.
 * `cats-algebra`: Type classes to represent algebraic structures.
 * `cats-free`: Free structures such as the free monad, and supporting type classes.
 * `cats-testkit`: lib for writing tests for type class instances using laws.
 * `alleycats-core`: Cats instances and classes which are not lawful. 
 
 There are several other Cats modules that are in separate repos so that they can 
 maintain independent release cycles. 
 
 * [`cats-effect`](https://github.com/typelevel/cats-effect): standard `IO` type together with `Sync`, `Async` and `Effect` type classes 
 * [`cats-mtl`](https://github.com/typelevel/cats-mtl): transformer typeclasses for Cats' Monads, Applicatives and Functors.
 * [`mouse`](https://github.com/typelevel/mouse): a small companion to Cats that provides convenient syntax (aka extension methods) 
 * [`kittens`](https://github.com/typelevel/kittens): automatic type class instance derivation for Cats and generic utility functions
 * [`cats-tagless`](https://github.com/typelevel/cats-tagless): Utilities for tagless final encoded algebras
 * [`cats-collections`](https://github.com/typelevel/cats-collections): Data structures which facilitate pure functional programming
 * [`cats-testkit-scalatest`](https://github.com/typelevel/cats-testkit-scalatest): Cats testkit integration with Scalatest

Past release notes for Cats are available in [CHANGES.md](https://github.com/typelevel/cats/blob/main/CHANGES.md).

### Documentation

Links:

1. Website: [typelevel.org/cats/](https://typelevel.org/cats/)
2. ScalaDoc: [typelevel.org/cats/api/](https://typelevel.org/cats/api/)
3. Type classes: [typelevel.org/cats/typeclasses](https://typelevel.org/cats/typeclasses.html)
4. Data types: [typelevel.org/cats/datatypes.html](https://typelevel.org/cats/datatypes.html)
5. Algebra overview: [typelevel.org/cats/algebra.html](https://typelevel.org/cats/algebra.html)
6. Glossary: [typelevel.org/cats/nomenclature.html](https://typelevel.org/cats/nomenclature.html)
7. Resources for Learners: [typelevel.org/cats/resources_for_learners.html](https://typelevel.org/cats/resources_for_learners.html)
8. FAQ: [typelevel.org/cats/faq.html](https://typelevel.org/cats/faq.html)
9. The Typelevel Ecosystem: [typelevel.org/cats/typelevelEcosystem.html](https://typelevel.org/cats/typelevelEcosystem.html)

### Community

Discussion around Cats is currently happening on Github issue and PR pages
as well as in two Gitter channels: 

[Gitter channel cats](https://gitter.im/typelevel/cats) is for general user 
questions and discussions, and 

[Gitter channel cats-dev](https://gitter.im/typelevel/cats-dev)
is dedicated for Cats development related discussions. For people who wants to 
follow closely and/or to participate in the decisions in Cats development, 
this is the room to join. 

People are expected to follow the
[Scala Code of Conduct](https://www.scala-lang.org/conduct/) when
discussing Cats on the Github page, Gitter channel, or other
venues.

We hope that our community will be respectful, helpful, and kind. If
you find yourself embroiled in a situation that becomes heated, or
that fails to live up to our expectations, you should disengage and
contact one of the [project maintainers](#maintainers) in private. We
hope to avoid letting minor aggressions and misunderstandings escalate
into larger problems.

If you are being harassed, please contact one of [us](#maintainers)
immediately so that we can support you.

### Binary compatibility and versioning

After `1.0.0` release, we [decided](https://github.com/typelevel/cats/issues/1233) 
to use *MAJOR.MINOR.PATCH* [Semantic Versioning 2.0.0](http://semver.org/)
going forward, which is different from the *EPOCH.MAJOR.MINOR* scheme common among 
Java and Scala libraries (including the Scala lang). 

Cats strives to provide a solid and stable foundation for an ecosystem of
FP libraries. Thus, we treat backward binary compatibility maintenance with a high priority. 
In semantic versioning, backward breaking change is **ONLY** allowed between *MAJOR* versions.
We will maintain backward binary compatibility between *PATCH* AND *MINOR* versions.
For example, when we release Cats `1.1.0`, it will be backward binary compatible 
with the previous `1.0.x` versions. I.E. the new JAR will be a drop-in replacement for 
the old one. This is critical when your application has a diamond
dependency on Cats - depending on two or more libraries that all depend on Cats. 
If one library upgrades to the new `1.1.0` Cats before the other one does, your 
application still runs thanks to this backward binary compatibility.

Also worth noting is that according to semantic versioning, 
*MINOR* version Y (x.Y.z | x > 0) MUST be incremented 
if new, backwards compatible functionality is introduced to the public API. 
It MUST be incremented if any public API functionality is marked as deprecated.

Any binary breaking changes will require a *MAJOR* version bump, which we will be very
cautious about. We will also consider using `organization` and package name for major 
versioning in the future. But that decision is yet to be made. 

### Adopters

Here's a (non-exhaustive) list of companies that use Cats in production. Don't see yours? [You can add it in a PR!](https://github.com/typelevel/cats/edit/main/README.md). And if you can, consider [supporting us](https://donorbox.org/typelevel-sustainability-program-2019?default_interval=m). 

- [Abacus Protocol](https://abacusprotocol.com)
- [Anduin Transactions](https://anduintransact.com)
- [Apple Inc. (FEAR team)](https://news.ycombinator.com/item?id=16969118)
- [AutoScout24](https://www.autoscout24.com) 
- [Avast](https://avast.com)
- [BabylonHealth](https://www.babylonhealth.com/)
- [Banno Group inside of Jack Henry & Associates](https://banno.com/)
- [Basefarm](https://basefarm.com/)
- [buildo](https://buildo.io)
- [Chartboost](https://www.chartboost.com/)
- [Cleverbase](https://cleverbase.com/en/)
- [Codacy](https://www.codacy.com/)
- [Codecentric](https://codecentric.de)
- [Colisweb](https://www.colisweb.com/)
- [Coya](https://coya.com/)
- [Datum Brain](https://datumbrain.com/)
- [DriveTribe](https://drivetribe.com/)
- [Dwolla](https://dwolla.com/)
- [Earnest](https://www.earnest.com)
- [eBay Inc.](https://www.ebay.com)
- [Eloquentix](https://eloquentix.com)
- [eSailors](https://www.esailors.de)
- [Evolution Gaming](https://www.evolutiongaming.com/)
- [Evotor Marketplace](https://market.evotor.ru/)
- [e.near](http://enear.co)
- [E.ON](https://eon.com)
- [formation.ai](https://formation.ai)
- [Free2Move](https://free2move.com)
- [HelloSoda](https://hellosoda.com)
- [HiFi](https://hi.fi)
- [HomeAway](https://www.homeaway.com)
- [Hootsuite Inc.](https://hootsuite.com/)
- [iHeartRadio](https://iheart.com)
- [ImmobilienScout24](https://www.immobilienscout24.de/)
- [Immozentral](https://www.immozentral.com/)
- [innFactory GmbH](https://www.innfactory.de/)
- [Iterators](https://www.iteratorshq.com/)
- [ITV](https://www.itv.com/)
- [Lightform](https://lightform.com/)
- [Lookout](https://www.lookout.com)
- [Merit](https://merits.com)

CHANGES.md
## Version 2.6.0

_2021 April 20_

Cats 2.6.0 is a binary-compatible release in the 2.x series. It provides support for Scala 2.12, 2.13, 3.0.0-RC2 and 3.0.0-RC3.

### Additions

* #3831 Add composition to Representable by @TimWSpence 
* #3832 Representable instance for Tuple2K by @TimWSpence 
* #3837 Provide instance of `Alternative[Set]` in alleycats by @rtyley 
* #3836 RepresentableStoreT by @TimWSpence 

### Changes

* #3847 Make deprecated catsDataInstancesForNonEmptyMap not implicit by @joroKr21 (source breaking)
* #3720, #3852 fix show interpolation ambiguity with Seq/List by @tnielens 
* #2254, #3856 Remove redundant Monad constraints from Parallel syntax by @joroKr21 
* #3851 Update 'Jump Start Guide' page to modern cats by @chaitanyawaikar

## Version 2.5.0

_2021 March 29_

Cats 2.5.0 is a binary-compatible release in the 2.x series. It provides support for Scala 2.12, 2.13, 3.0.0-RC1 and 3.0.0-RC2.

### Additions

* #3808 Chain add fromOption factory method by @nikiforo 
* #3826, #3827 Add syntax for parProduct and parAp by @nigredo-tori 
* #3804, #3805 Add Ior.mergeWith and AlignMergeWith by @nigredo-tori 
* #3806 Add Align instance for Nested by @nigredo-tori 
* #3643, #3823 Add sumAll & productAll by @phongngtuan, @larsrh 

### Changes

* #3818 update Scala to 2.13.5 by @sh0hei 
* #3825 Use -scalajs-mapSourceURI in Scala 3 builds by @fthomas 
* #3377, #3807 Replace Band instance for NonEmptyMap with a Semigroup instance consistent with SortedMap by @nigredo-tori 
* #3480, #3833 Rename master to main by @travisbrown, @larsrh 

#### Documentation

* #3814 typo in `ApplicativeError` by @wibisono 

## Version 2.4.1

_2021 February 10_

This release fixes a problem in Cats 2.4.0. It is not binary compatible with Cats 2.4.0. We urge everyone to skip 2.4.0 and upgrade directly to 2.4.1.

Cats 2.4.1 is a hotfix release in the 2.x series. It is binary compatible to every previous release in the 2.x series, _except for Cats 2.4.0_.

The following release notes contain all cumulative changes since Cats 2.3.1.

### Additions

* #3220, #3222, #3773 Add Invariant instances for Numeric and Integral by @tmccarthy & @kubukoz 
* #3650, #3579 add grouped to NonEmptyCollection by @bbstilson & @larsrh 
* #3299, #3293 Generate Show instances for Tuple3 and higher by @barambani 
* #3727 .slidingN using tuples by @Slakah 
* #3750, #1549 Support scala-native 0.4 by @arashi01 
* #3771 add NonEmptyChain.of by @larsrh 
* #3318 add syntax for unzip on functors by @yangzai 
* #3680 Add more group-like methods to NonEmptyList & Chain by @BalmungSan 
* #3748 Delimited continuations via shiftT/resetT for ContT by @TimWSpence 
* #3747 Add callCC to ContT by @TimWSpence 
* #3765 Add NonEmptyList.take by @saeltz 
* #2087, #3753 add Monoid[Comparison] for lexographical ordering by @adelbertc & @larsrh 
* #3759 Provide BoundedEnumerable[Byte] instance by @ashleymercer 
* #3699 Allow finding a list of maxiumum/minimum items for Foldable/Reducible by @Billzabob & @larsrh 
* #3730 Add NonEmptyList.zip by @saeltz 
* #3724 Trivial semigroup instances by @RaasAhsan 
* #3707, #3710 Add missing Free.liftK and Free.liftId by @phongngtuan 

### Changes

* #3776, #3777 Removed and replaced inconsistent Parallel derivation for EitherT by @djspiewak 
  * this change is potentially source-breaking: an existing `implicit` has been made un-`implicit`, but another `implicit` with a different semantics has been provided (core)
  * more discussion at typelevel/cats-effect#1645
  * this change emits new deprecation warnings
* #3734, #3772 Cogen for (Finite)Duration to fix test failures by @larsrh 
  * this change is only internal in tests
* #2207, #3758 deprecate Free.inject by @larsrh 
  * this change emits new deprecation warnings
* #3702 specialized EqOps by @YannMoisan 
  * this change accidentally broke source compatibility, please upgrade to 2.4.1
  * reverted by #3784, #3785 by @larsrh 
* #3677 Added EmptyK instance for Map by @LMnet 
  * this change marks a previously non-`implicit` method as `implicit` (alleycats)

#### Documentation

* #3722 Rename productM to mproduct in the glossary by @BenFradet 
* #3745 Adjust to GitHub Actions by @sh0hei 
* #3744 Improve scaladoc for EitherT.rethrowT by @majk-p 
* #3732 scaladoc fixes for PartialOrder by @martijnhoekstra 
* #3728 Fix a link in typelevelEcosystem.md by @akiomik 
* #3726 update README by @gokyo 

## Version 2.4.0

_2021 February 9_

Cats 2.4.0 contains an unintended source-breaking change, please upgrade directly to 2.4.1.

## Version 2.3.1

_2020 December 18_

Cats 2.3.1 adds support for Scala 3.0.0-M3 and is published for Scala 2.12, 2.13, and 3.0.0-M2 as well.

### Changes From 2.3.0

- #3696 Add `ApplicativeThrow` and `MonadThrow` to cats package object (from cats-effect) (@agustafson)
- #3698 Add addLeft/addRight for Ior (@keirlawson)
- #3639 Created documentation for monaderror, merged it with applicative error (@dhinojosa)
- #3402 Simplify FunctionK.lift macro (@joroKr21)
- #3708 Use as rather than map and discard (@johnynek)

## Version 2.3.0

_2020 November 26_

Welcome to the first stable release of Cats on Scala 3… milestone 2! We will continue releasing minor versions within the 2.3.x lineage against Scala 3.0.0 until it reaches a final stable release. The decision to push a stable release of Cats in advance of this date was made in order to simplify versioning within the downstream ecosystem.

### Changes From 2.2.0

- #3673 – tweak Band impl so it can more easily be combined with Monoid in a single instance (@Dwolla)
- #3609 – A new dedicated cats ecosystem page (@benkio)
- #3630 – FunctionK: methods to narrow input / widen output types (@bplommer)
- #3620, #3624 – Typeclass instances for `immutable.Seq` (@JosephMoniz)
- #3603 – Add method `distinct` to TraverseFilter (@takayahilton)

## Version 2.3.0-M2

_2020 November 6_ 🍂

Added support for Scala/Dotty 3.0.0-M1.

### Other improvements

* #3661 fix typo in Scaladoc by @fixxer 

project/Unidoc.scala
package sbtunidoc

import sbt._
import Keys._

object Plugin extends sbt.Plugin {
  lazy val JavaUnidoc = config("javaunidoc") extend Compile
  lazy val ScalaUnidoc = config("scalaunidoc") extend Compile
  lazy val TestJavaUnidoc = config("testjavaunidoc") extend Test
  lazy val TestScalaUnidoc = config("testscalaunidoc") extend Test
  lazy val Genjavadoc = config("genjavadoc") extend Compile

  import UnidocKeys._

  object UnidocKeys {
    val unidoc                    = taskKey[Seq[File]]("Create unified scaladoc for all aggregates.")
    val unidocAllSources          = taskKey[Seq[Seq[File]]]("All sources.")
    val unidocAllClasspaths       = taskKey[Seq[Classpath]]("All classpaths.")
    val unidocAllAPIMappings      = taskKey[Seq[Map[File, URL]]]("All API mappings.")
    val unidocScopeFilter         = settingKey[ScopeFilter]("Control sources to be included in unidoc.")
    val unidocProjectFilter       = settingKey[ScopeFilter.ProjectFilter]("Control projects to be included in unidoc.")
    val unidocConfigurationFilter = settingKey[ScopeFilter.ConfigurationFilter]("Control configurations to be included in unidoc.")
    val unidocGenjavadocVersion   = settingKey[String]("Version of the genjavadoc compiler plugin.")
  }

  def baseCommonUnidocTasks(sc: Configuration): Seq[sbt.Def.Setting[_]] = Seq(
    doc := Unidoc(streams.value.cacheDirectory, (compilers in unidoc).value, (sources in unidoc).value, (fullClasspath in unidoc).value,
      (scalacOptions in unidoc).value, (javacOptions in unidoc).value, (apiMappings in unidoc).value, (maxErrors in unidoc).value,
      (target in unidoc).value, configuration.value, streams.value),
    compilers in unidoc := (compilers in sc).value,
    sources in unidoc := (unidocAllSources in unidoc).value.flatten,
    scalacOptions in unidoc := (scalacOptions in (sc, doc)).value,
    javacOptions in unidoc := (javacOptions in (sc, doc)).value,
    fullClasspath in unidoc := (unidocAllClasspaths in unidoc).value.flatten.distinct,
    unidocAllClasspaths in unidoc := Unidoc.allClasspathsTask.value,
    apiMappings in unidoc := {
      val all = (unidocAllAPIMappings in unidoc).value
      val allList = all map { _.toList }
      allList.flatten.distinct.toMap
    },
    unidocAllAPIMappings in unidoc := Unidoc.allAPIMappingsTask.value,
    maxErrors in unidoc := (maxErrors in (sc, doc)).value,
    unidocScopeFilter in unidoc := ScopeFilter((unidocProjectFilter in unidoc).value, (unidocConfigurationFilter in unidoc).value),
    unidocProjectFilter in unidoc := inAnyProject,
    // {
    //   val exclude = excludedProjects.value
    //   inAnyProject -- inProjects(buildStructure.value.allProjectRefs filter { p => exclude contains (p.project) }: _*)
    // },
    // excludedProjects in unidoc := Seq(),
    unidocConfigurationFilter in unidoc := inConfigurations(sc),
    unidocGenjavadocVersion in Global := "0.10"
  )
  def baseScalaUnidocTasks(sc: Configuration): Seq[sbt.Def.Setting[_]] = baseCommonUnidocTasks(sc) ++ Seq(
    target in unidoc := crossTarget.value / "unidoc",
    unidocAllSources in unidoc := Unidoc.allScalaSources.value
  )
  def baseJavaUnidocTasks(sc: Configuration): Seq[sbt.Def.Setting[_]] = baseCommonUnidocTasks(sc) ++ Seq(
    target in unidoc := target.value / "javaunidoc",
    unidocAllSources in unidoc := Unidoc.allJavaSourcesTask.value
  )
  def baseGenjavadocExtraTasks(sc: Configuration): Seq[sbt.Def.Setting[_]] = Seq(
    artifactName in packageDoc := { (sv, mod, art) => "" + mod.name + "_" + sv.binary + "-" + mod.revision + "-javadoc.jar" },
    sources <<= (target, sources in sc, compile in sc) map { (t, s, c) =>
      (t / "java" ** "*.java").get ++ s.filter(_.getName.endsWith(".java"))
    },
    javacOptions in doc := (javacOptions in (sc, doc)).value
  )
  /** Add this to child projects to generate equivalent java files out of scala files. */
  lazy val genjavadocSettings: Seq[sbt.Def.Setting[_]] = Seq(
    libraryDependencies += compilerPlugin("com.typesafe.genjavadoc" %% "genjavadoc-plugin" % unidocGenjavadocVersion.value cross CrossVersion.full),
    scalacOptions <+= target map (t => "-P:genjavadoc:out=" + (t / "java")))
  /** Add this to child projects to replace packaged javadoc with the genjavadoc. */
  lazy val genjavadocExtraSettings: Seq[sbt.Def.Setting[_]] = genjavadocExtraTask(Genjavadoc, Compile)
  def genjavadocExtraTask(c: Configuration, sc: Configuration): Seq[sbt.Def.Setting[_]] =
    genjavadocSettings ++
    inConfig(c)(Defaults.configSettings ++ baseGenjavadocExtraTasks(sc)) ++ Seq(
      packageDoc in sc <<= packageDoc in c
    )

  /** Add this to the root project to generate Java unidoc. */
  lazy val javaUnidocSettings: Seq[sbt.Def.Setting[_]] =
    javaUnidocTask(JavaUnidoc, Compile) ++
    javaUnidocTask(TestJavaUnidoc, Test) ++
    inConfig(TestJavaUnidoc)(Seq(
      target in unidoc := target.value / "testjavaunidoc"
    ))
  def javaUnidocTask(c: Configuration, sc: Configuration): Seq[sbt.Def.Setting[_]] =
    inConfig(c)(Defaults.configSettings ++ baseJavaUnidocTasks(sc)) ++ Seq(
      unidoc in sc := Seq((doc in c).value)
    )

  /** Add this to the root project to generate Scala unidoc. */
  lazy val scalaUnidocSettings: Seq[sbt.Def.Setting[_]] =
    scalaUnidocTask(ScalaUnidoc, Compile) ++
    scalaUnidocTask(TestScalaUnidoc, Test) ++
    inConfig(TestScalaUnidoc)(Seq(
      target in unidoc := crossTarget.value / "testunidoc"
    ))
  def scalaUnidocTask(c: Configuration, sc: Configuration): Seq[sbt.Def.Setting[_]] =
    inConfig(c)(Defaults.configSettings ++ baseScalaUnidocTasks(sc)) ++ Seq(
      unidoc in sc := Seq((doc in c).value)
    )

  /** An alias for scalaUnidocSettings */
  lazy val unidocSettings: Seq[sbt.Def.Setting[_]] = scalaUnidocSettings
  /** Add this to the root project to generate both Scala unidoc and Java unidoc. */
  lazy val scalaJavaUnidocSettings: Seq[sbt.Def.Setting[_]] =
    scalaJavaUnidocTask(ScalaUnidoc, JavaUnidoc, Compile) ++
    scalaJavaUnidocTask(TestScalaUnidoc, TestJavaUnidoc, Test)
  def scalaJavaUnidocTask(c1: Configuration, c2: Configuration, sc: Configuration): Seq[sbt.Def.Setting[_]] =
    inConfig(c1)(Defaults.configSettings ++ baseScalaUnidocTasks(sc)) ++
    inConfig(c2)(Defaults.configSettings ++ baseJavaUnidocTasks(sc)) ++ Seq(
      unidoc in sc <<= (doc in c1, doc in c2) map { (s, j) => Seq(s, j) })

  object Unidoc {
    import java.io.PrintWriter

    // This is straight out of docTaskSettings in Defaults.scala.
    def apply(cache: File, cs: Compiler.Compilers, srcs: Seq[File], cp: Classpath,
      sOpts: Seq[String], jOpts: Seq[String], xapis: Map[File, URL], maxErrors: Int,
      out: File, config: Configuration, s: TaskStreams): File = {
      val hasScala = srcs.exists(_.name.endsWith(".scala"))
      val hasJava = srcs.exists(_.name.endsWith(".java"))
      val label = nameForSrc(config.name)
      val (options, runDoc) =
        if(hasScala)
          (sOpts ++ Opts.doc.externalAPI(xapis), // can't put the .value calls directly here until 2.10.2
           // InTheNow: changed Doc.scaladoc to scaladoc
            scaladoc(label, cache / "scala", cs.scalac.onArgs(exported(s, "scaladoc"))))
        else if(hasJava)
          (jOpts,
            Doc.javadoc(label, cache / "java", cs.javac.onArgs(exported(s, "javadoc"))))
        else
          (Nil, RawCompileLike.nop)
      runDoc(srcs, cp map {_.data}, out, options, maxErrors, s.log)
      out
    }

    private[this] def exported(w: PrintWriter, command: String): Seq[String] => Unit = args =>
      w.println( (command +: args).mkString(" ") )
    private[this] def exported(s: TaskStreams, command: String): Seq[String] => Unit = args =>
      exported(s.text("export"), command)
    def nameForSrc(name: String): String = name match {
      case "compile"|"javaunidoc"|"scalaunidoc" => "main"
      case _ => name
    }
    lazy val allScalaSources = Def.taskDyn {
      val f = (unidocScopeFilter in unidoc).value
      sources.all(f)
    }
    lazy val javaSources: sbt.Def.Initialize[Task[Seq[File]]] = Def.task {
      val compiled = compile.value
      val sourceJavaFiles = sources.value filter {_.getName endsWith ".java"}
      val targetJavaFiles: Seq[File] = (target.value / "java" ** "*.java").get.sorted
      sourceJavaFiles ++ targetJavaFiles
    }
    lazy val allJavaSourcesTask = Def.taskDyn {
      val f = (unidocScopeFilter in unidoc).value
      javaSources.all(f)
    }
    lazy val allClasspathsTask = Def.taskDyn {
      val f = (unidocScopeFilter in unidoc).value
      dependencyClasspath.all(f)
    }
    lazy val allAPIMappingsTask = Def.taskDyn {
      val f = (unidocScopeFilter in unidoc).value
      (apiMappings in (Compile, doc)).all(f)
    }

     // InTheNow: Added rest of code
    /*
      There are errors when generating scaladoc in cats (and other projects, eg circe) in 2.10. Up to now,
      these projects have assumed the issue was a "2.10" problem, and just disabled docs for 2.10.
      However, cats cannot also generate its laws scaladoc in 2.11. The tail output is:

```
[error] ./laws/src/main/scala/cats/laws/ReducibleLaws.scala:29: value sequence1_ is not a member of type parameter F[G[A]]
[error]     fa.sequence1_ <-> fa.sequence_
[error]        ^
[error] ./laws/src/main/scala/cats/laws/TraverseLaws.scala:67: value foldMap is not a member of type parameter F[A]
[error]     val rhs: B = fa.foldMap(f)
[error]                     ^
[info] No documentation generated with unsuccessful compiler run
[error] 28 errors found
[info] Main Scala API documentation successful.
```

      Note that no documentation was generated, but the last "info" states ""documentation successful""

      So I investigated:

      Running scaladoc directly from the command line yields:

```
./laws/src/main/scala/cats/laws/ReducibleLaws.scala:29: error: value sequence1_ is not a member of type parameter F[G[A]]
    fa.sequence1_ <-> fa.sequence_
       ^                            ^
./laws/src/main/scala/cats/laws/TraverseLaws.scala:67: error: value foldMap is not a member of type parameter F[A]
    val rhs: B = fa.foldMap(f)
                    ^
model contains 738 documentable templates
28 errors found
```

     Note same errors, different formatting - but documentation is generated.

     So, I believe the differnce in behaviour is down to that scaladoc uses its own reporter,
     that essentially ignores errors.

     see https://github.com/scala/scala/blob/2.12.x/src/scaladoc/scala/tools/nsc/ScalaDoc.scala#L61-L64

     So the change here is to use a custom reporter, that also ignores errrors. I have also changed the output to
     print the message with "error :" (as in the commandline version), as turn the error to a warning. This way,
     it looks less weird.

     Possibly, we could adapt scaladoc in the same way, adding to
     https://github.com/scala/scala/pull/4032/commits/5f29a264469b6ae161d9a98fe76d3f6466d82b12
     */
    import compiler.{ AnalyzingCompiler}
    import RawCompileLike._
    import xsbti.{Position, Severity}
    import Severity.{ Error, Warn }

    private[this] def scaladoc(label: String, cache: File, compiler: AnalyzingCompiler): Gen =
      scaladoc(label, cache, compiler, Seq())

    private[this] def scaladoc(label: String, cache: File, compiler: AnalyzingCompiler, fileInputOptions: Seq[String]): Gen = {
      def doc(sources: Seq[File], classpath: Seq[File], outputDirectory: File, options: Seq[String], maximumErrors: Int, log: Logger): Unit =
        compiler.doc(sources, classpath, outputDirectory, options, log, new ScaladocReporter(maximumErrors, log))

      cached(cache, fileInputOptions, prepare(label + " Scala API documentation", doc))
    }

    class ScaladocReporter (maximumErrors: Int, log: Logger) extends sbt.LoggerReporter(maximumErrors, log) {
      override def hasErrors = false

      override def log(pos: Position, msg: String, severity: Severity): Unit = {
        val newSeverity = if (severity == Error) Warn else severity
        super.log(pos, s"error: $msg", newSeverity )
      }
    }
  }
}
